onnxruntime/core/providers/qnn/qnn_allocator.cc

```diff
@@ -6,91 +6,94 @@
 #include <cassert>
 #include <cstddef>
 #include <algorithm>
+#include <shared_mutex>
 
 #include "core/providers/qnn/ort_api.h"
 
 namespace onnxruntime::qnn {
 
-/**
- * HtpSharedMemoryAllocator allocation details
- *
- * The HTP shared memory allocator will allocate a block of shared memory larger than the amount requested in order to
- * hold some additional info.
- * Each allocation returned by HtpSharedMemoryAllocator::Alloc() is preceded by an AllocationHeader structure.
- *
- * For example, if Alloc(num_requested_bytes) is called, this is what the memory layout looks like:
- *   | AllocationHeader bytes | num_requested_bytes bytes |
- *                              ^- address returned by Alloc()
- *
- * The AllocationHeader can be used to obtain the owning allocator instance, which in turn can be used to do other
- * operations with that allocation, such as retrieving more info about the allocation.
- */
-
 namespace {
 
-struct AllocationHeader {
-  static constexpr std::array<char, 8> kAllocationHeaderMarker{'o', 'r', 't', 'a', 'l', 'l', 'o', 'c'};
-
-  // Marker bytes to verify as a sanity check.
-  std::array<char, 8> marker;
-
-  // Pointer to the allocating allocator instance.
-  // Note: A critical assumption here is that the allocating allocator is not destroyed before the allocation is freed.
-  HtpSharedMemoryAllocator* allocator_ptr;
-
-  AllocationHeader(HtpSharedMemoryAllocator* allocator_ptr)
-      : marker{kAllocationHeaderMarker},
-        allocator_ptr{allocator_ptr} {
-  }
-
-  ~AllocationHeader() {
-    marker.fill('\0');
-    allocator_ptr = nullptr;
-  }
-};
-
 size_t AllocationAlignment() {
   constexpr size_t min_allocation_alignment = 64;  // Equal to MlasGetPreferredBufferAlignment()
-  return std::max(alignof(AllocationHeader), min_allocation_alignment);
-}
-
-size_t DivRoundUp(size_t a, size_t b) {  // TODO is there already a helper function somewhere for this?
-  return (a + b - 1) / b;
+  return min_allocation_alignment;
 }
 
 bool IsAligned(const void* address, size_t alignment) {
   assert((alignment & (alignment - 1)) == 0);  // alignment must be a power of two
   return (reinterpret_cast<uintptr_t>(address) & (alignment - 1)) == 0;
 }
 
-size_t AllocationOffsetFromStartOfHeader() {
-  const size_t allocation_alignment = AllocationAlignment();
-  const size_t offset = DivRoundUp(sizeof(AllocationHeader), allocation_alignment) * allocation_alignment;
-  return offset;
+std::unique_ptr<void, void (*)(void*)> WrapSharedMemoryWithUniquePtr(void* shared_memory_raw,
+                                                                     const RpcMemApi& rpcmem_api) {
+  return {shared_memory_raw, rpcmem_api.free};
 }
 
-std::byte* GetAllocationHeaderAddress(void* allocation_address) {
-  auto* allocation_header_address = reinterpret_cast<std::byte*>(allocation_address) - sizeof(AllocationHeader);
-  return allocation_header_address;
+class AllocationTracker {
```

Comment:
@edgchen1 could you take a look over these changes as I believe you added the original implementation?

Reply:
I also made these changes. I'll give them a look but it would be good to have more eyes on them too.